src/interchange/src/protobuf.rs

```diff
@@ -228,98 +184,67 @@ fn derive_inner_type(
-        ReflectFieldRef::Optional(None) => {
-            if field_desc.get_proto().get_label() == Label::LABEL_REQUIRED {
+fn pack_message(packer: &mut Row, message: &DynamicMessage) -> Result<(), anyhow::Error> {
+    for field_desc in message.descriptor().fields() {
+        if !message.has_field(&field_desc) {
+            if field_desc.cardinality() == Cardinality::Required {
                 bail!(
                     "protobuf message missing required field {}",
-                    field_desc.get_name()
+                    field_desc.name()
                 );
             }
-            match field_desc.singular_runtime_type() {
-                RuntimeTypeBox::Message(_) => packer.push(Datum::Null),
-                _ => pack_value(packer, field_desc, field_desc.singular_default_value())?,
+            if field_desc.kind().as_message().is_some() && !field_desc.is_list() {
+                packer.push(Datum::Null);
+                continue;
             }
         }
-        ReflectFieldRef::Optional(Some(value)) => pack_value(packer, field_desc, value)?,
-        ReflectFieldRef::Repeated(values) => packer.push_list_with(|packer| {
-            for value in values {
-                pack_value(packer, field_desc, value)?;
-            }
-            Ok::<_, anyhow::Error>(())
-        })?,
-        ReflectFieldRef::Map(_) => {
-            bail!("internal error: unexpected map field while decoding protobuf")
-        }
+        let value = message.get_field(&field_desc);
+        pack_value(packer, &field_desc, &*value)?;
     }
     Ok(())
 }
 
 fn pack_value(
     packer: &mut Row,
     field_desc: &FieldDescriptor,
-    value: ReflectValueRef,
+    value: &Value,
 ) -> Result<(), anyhow::Error> {
     match value {
-        ReflectValueRef::Bool(false) => packer.push(Datum::False),
-        ReflectValueRef::Bool(true) => packer.push(Datum::True),
-        ReflectValueRef::I32(i) => packer.push(Datum::Int32(i)),
-        ReflectValueRef::I64(i) => packer.push(Datum::Int64(i)),
-        ReflectValueRef::F32(f) => packer.push(Datum::Float32(f.into())),
-        ReflectValueRef::F64(f) => packer.push(Datum::Float64(f.into())),
-        ReflectValueRef::String(s) => packer.push(Datum::String(s)),
-        ReflectValueRef::Bytes(s) => packer.push(Datum::Bytes(s)),
-        ReflectValueRef::Enum(enum_desc, i) => match enum_desc.get_value_by_number(i) {
+        Value::Bool(false) => packer.push(Datum::False),
+        Value::Bool(true) => packer.push(Datum::True),
+        Value::I32(i) => packer.push(Datum::Int32(*i)),
+        Value::I64(i) => packer.push(Datum::Int64(*i)),
+        Value::F32(f) => packer.push(Datum::Float32((*f).into())),
+        Value::F64(f) => packer.push(Datum::Float64((*f).into())),
+        Value::String(s) => packer.push(Datum::String(s)),
+        Value::Bytes(s) => packer.push(Datum::Bytes(s)),
+        Value::EnumNumber(i) => match field_desc.kind().as_enum().and_then(|e| e.get_value(*i)) {
```

Comment:
is this basically asserting that the as_enum() will never return None? otherwise the bail! message could be inaccurate

Reply:
True. Updated to yield two different error messages.